src/main/scala/pureconfig/ConfigConvert.scala

```diff
@@ -6,227 +6,199 @@
-  }
-
-  // traversable of simple types, that are types with an instance of StringConvert
-  implicit def deriveTraversable[T, F[T] <: TraversableOnce[T]](implicit stringConvert: Lazy[StringConvert[T]],
-    cbf: CanBuildFrom[F[T], T, F[T]]) = new ConfigConvert[F[T]] {
-    override def from(config: RawConfig, namespace: String): Try[F[T]] = {
-      val value = config(namespace)
-
-      if (value.isEmpty) {
-        // "".split(",") returns Array("") but we want Array()
-        Success(cbf().result())
-      } else {
-        val tryBuilder = value.split(",").foldLeft(Try(cbf())) {
-          case (tryResult, rawValue) =>
-            for {
-              result <- tryResult
-              value <- stringConvert.value.from(rawValue)
-            } yield result += value
-        }
-
-        tryBuilder.map(_.result())
-      }
+    override def to(t: Option[T]): ConfigValue = t match {
+      case Some(v) => conv.value.to(v)
+      case None => ConfigValueFactory.fromMap(Map().asJava)
     }
 
-    override def to(ts: F[T], namespace: String): RawConfig =
-      Map(namespace -> ts.map(stringConvert.value.to).mkString(","))
+    def toOption(t: Option[T]): Option[ConfigValue] = t.map(conv.value.to)
   }
 
-  // traversable of complex types, that are types with an instance of ConfigConvert
-  implicit def deriveTraversableOfObjects[T, F[T] <: TraversableOnce[T]](implicit configConvert: Lazy[ConfigConvert[T]],
+  // traversable of types with an instance of ConfigConvert
+  implicit def deriveTraversable[T, F[T] <: TraversableOnce[T]](implicit configConvert: Lazy[ConfigConvert[T]],
     cbf: CanBuildFrom[F[T], T, F[T]]) = new ConfigConvert[F[T]] {
 
-    override def from(config: RawConfig, namespace: String): Try[F[T]] = {
-      val fullKeysFound = config.keys.filter(_ startsWith (namespace + namespaceSep)).toList
-      val keysFound = fullKeysFound.map { key =>
-        val parentNamespaceLength = namespace.length + namespaceSep.length
-        key.substring(0, key.indexOfSlice(namespaceSep, parentNamespaceLength))
-      }.sorted.distinct
-
-      if (keysFound.isEmpty) {
-        Success(cbf().result())
-      } else {
-        val tryBuilder = keysFound.foldLeft(Try(cbf())) {
-          case (tryResult, key) =>
-            for {
-              result <- tryResult
-              value <- configConvert.value.from(config, key)
-            } yield result += value
-        }
-
-        tryBuilder.map(_.result())
+    override def from(config: ConfigValue): Try[F[T]] = {
+      config match {
+        case co: ConfigList =>
+          val tryBuilder = co.asScala.foldLeft(Try(cbf())) {
+            case (tryResult, v) =>
+              for {
+                result <- tryResult
+                value <- configConvert.value.from(v)
+              } yield result += value
+          }
+
+          tryBuilder.map(_.result())
+        case _ =>
+          Success(cbf().result())
```

Comment:
I'm not sure about this: if `config` is not a `ConfigList`, we ignore it and return an empty traversable while probably that's not what we want because in this way. The `case _` should be an error unless we want to consider `null` as empty traversable. But in all the other cases, this should be an error saying that the type of `config` cannot be converted to a traversable.